Refactor ESP32xx folder structure

marekmatej opened the pull request:
This PR aims to refactor Espressif `soc` folders for all architectures, in order to match current Zephyrs standards and to better reflect ESP32 product line SoC and SiP.

marekmatej committed:
dts: riscv: esp32c3 rework soc/sip list

Introduce dtsi files representing the
current portfolio of chips and modules
based on the:

- flash size
- psram size
- gpio count
- certification status

Update the boards dts files according
to which SOC/SIP they are using.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
dts: xtensa: esp32xx rework soc/sip list

Introduce dtsi files representing the
current portfolio of chips and modules
based on the followint criteria:

- flash size
- psram size
- gpio count
- certification status

Update the boards dts files according
to which SOC/SIP they are using.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
drivers: can: fix long lines

Change the line length to comply with the rules.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

LucasTambor commented on boards/xtensa/esp32_net/esp32_net_defconfig:
commented code

LucasTambor commented on modules/Kconfig.esp32:
```diff
@@ -3,4 +3,4 @@
 config HAS_ESPRESSIF_HAL
 	bool
-	depends on SOC_ESP32 || SOC_ESP32S2 || SOC_ESP32C3 || SOC_ESP32_NET || SOC_ESP32S3
+	depends on SOC_SERIES_ESP32 || SOC_SERIES_ESP32S2 || SOC_SERIES_ESP32S3 || SOC_SERIES_ESP32C3 || SOC_SERIES_ESP32_NET
```
Isn't it better to use SOC_FAMILY_ESP32?

LucasTambor commented on tests/drivers/uart/uart_async_api/src/test_uart.h:
```diff
@@ -21,6 +21,8 @@
 
 #if DT_NODE_EXISTS(DT_NODELABEL(dut))
 #define UART_NODE DT_NODELABEL(dut)
+#elif defined(CONFIG_SOC_SERIES_ESP32C3)
```
ESP32C3 already have `dut` node in its overlay, right?

marekmatej replied:
yes, but not all the ESP32C3 boards has its overlay

LucasTambor replied:
Oh right. Then I think it is better to include an overlay for the target board with `dut` node if it wants to run this test. If you take a look at the file history, you will see that used to be a lot of `#ifdefs` defining the UART node according to the SOC until `dut` node was introduced. But I don't think including any overlay is in the scope of this PR.

marekmatej replied:
maybe we can update this files in upgoing PR, what do you think?

marekmatej committed:
include: dt-bingings: Fix typo

Fix minor typo in esp32s3-gpio-sigmap

Signed-off-by: Marek Matej <marek.matej@espressif.com>

henrikbrixandersen reviewed (commented):
CAN changes okay.

mniestroj commented on samples/arch/smp/pktqueue/sample.yaml:
```diff
@@ -19,7 +19,7 @@ tests:
     filter: (CONFIG_MP_MAX_NUM_CPUS > 1)
     platform_exclude:
-      - esp32
+      - esp32_devkitc_wroom
```
Such changes should be done together with the commit that split `esp32` into two separate platforms.

marekmatej replied:
Answered that in the later comment.

mniestroj commented on tests/boards/espressif_esp32/cache_coex/README.rst:
```diff
@@ -19,7 +19,7 @@ Make sure you have the ESP32 DevKitC connected over USB port.
 .. code-block:: console
 
-   west build -b esp32 tests/boards/espressif_esp32/cache_coex
+   west build -b esp32_devkitc_wrover tests/boards/espressif_esp32/cache_coex
```
Such changes should be done together with the commit that split `esp32` into two separate platforms.

marekmatej replied:
As much I agree with you, it is quite hard to achieve good code reviewability and proper bisectionability at the same time with this type of braking changes bringing PRs. So I choose to group the commits by the folder they alter. 
However, I agree on reordering them so the sooner commits bring changes/files for the later one.

mniestroj replied:
> As much I agree with you, it is quite hard to achieve good code reviewability and proper bisectionability at the same time with this type of braking changes bringing PRs.

I don't agree here regarding bisectability/reviewability. IMO bisectability comes in pair with reviewability. If logical changes are kept minimal, then both the diff and the portion of code to be reviewed are minimal. In this PR however the logical changes (like renaming `CONFIG_SOC` to `CONFIG_SOC_SERIES`) are split across multiple commits. This fails to build (which means broken bisectability) and reviewing commit by commit is not applicable as well (all the commits need to be inspected in order to see if the change was complete).

> So I choose to group the commits by the folder they alter.

This is actually something that comes for free, without splitting commits. If you look at commit diff, then updated code chunks are sorted by the directory/file (file path). So grouping commits by folder gives no added value.

marekmatej replied:
I've checked the 'Contributor expectations' document under the 'Contribution guidelines' and you are right about that one! Will fix that.

henrikbrixandersen replied:
Breaking bisectability will affect a lot of developers, please do do not do that.

marekmatej commented on boards/xtensa/esp32_devkitc_wroom/esp32_devkitc_wroom_defconfig:
```diff
@@ -0,0 +1,13 @@
+CONFIG_SERIAL=y
+CONFIG_UART_CONSOLE=y
+
+CONFIG_GPIO=y
```
The configs you are probably missing in `defconfig` were moved to a more proper destination in the `soc/xtensa/espressif_esp32/common/Kconfig.defconfig.series`. Those configs were blindly copied in every single ESP32 board without any changes.
I will emphasize this in the board altering commit message.

marekmatej commented on soc/xtensa/espressif_esp32/esp32/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,69 @@
+	default "esp32_wrover_e_n16r2"	if SOC_ESP32_WROVER_E_N16R2
+	default "esp32_wrover_e_n4r8"	if SOC_ESP32_WROVER_E_N4R8
+	default "esp32_wrover_e_n8r8"	if SOC_ESP32_WROVER_E_N8R8
+	default "esp32_wrover_e_n16r8"	if SOC_ESP32_WROVER_E_N16R8
```
Well, the modules `esp32_wroom`, `esp32_wrover` etc. have some substantial differences, like flash/psram size and also the GPIO availability (which is mostly handled by hal/IDF). And regarding the mixing modules with SOCs, I took inspiration from the already supported SOC/SIP. For example, `NRF9160` or `NRF9161` is listed the same way as other SOCs. 

Overall, this list of SOC/SIP comes from the ESP32 product selector, and I choose the up-to-date modules/soc or modules/soc which are already used by the board supported in Zephyr. Also, the list is meant to serve as a reference for any new board designers.

mniestroj replied:
> Well, the modules esp32_wroom, esp32_wrover etc. have some substantial differences, like flash/psram size and also the GPIO availability (which is mostly handled by hal/IDF).

In added dtsi files I see only flash/ram differences. Also, as an example, `esp32s2_mini_n4r2.dtsi`, `esp32s2_solo_n4r2.dtsi` and `esp32s2_wrover_n4r2.dtsi` are all the same. So is it worth introducing all of them? With all added features, there is a ratio of gain and increased maintenance. I see that those can be used by downstream boards, but is it enough to justify adding so many module variants and maintaining them in Zephyr upstream?

There also comes the question where to put those files, maybe it would be good to create some "module" directory to group all of them and separate them from SoC definitions.

> And regarding the mixing modules with SOCs, I took inspiration from the already supported SOC/SIP. For example, NRF9160 or NRF9161 is listed the same way as other SOCs.

nRF9160/nRF9161 is a bit different, as those can be obtained only in SiP, never in smaller entities. Each of them is probably based on different SoC, since they have different capabilities. ESP32 modules are something built on top of SoCs, they are clearly something between an SoC and a board (PCB of a ready product or devkit).

Can we move it to another PR, so we can progress with refactoring/bugfixes regardless of what we decide to do with modules?

marekmatej replied:
Regarding the SOC/SIP granularity - I created only those modules which have different properties. The `FLASH` size, `PSRAM` size, `GPIO` count, and the `certification` status were taken into account when making the selection. The selection wasn't random or bulk. Currently, not all properties are handled in dtsi, but will be added later in another PR (GPIO count).

Regarding the SIP vs SOC - SOCs used inside the SIPs are often different from the plain SOC by how they incorporate FLASH or PSRAM memories, which are often integrated into the same package/die, or how many GPIOs they offer for users. So the difference between SIP and SOC is from very blurry to nonexistent. So comparing `ESP32xx` modules to `NRF916x` is spot on.

marekmatej commented on drivers/adc/adc_esp32.c:
```diff
@@ -20,7 +20,7 @@
 LOG_MODULE_REGISTER(adc_esp32, CONFIG_ADC_LOG_LEVEL);
 
-#if CONFIG_SOC_ESP32
+#if CONFIG_SOC_SERIES_ESP32
```
No, the symbols are not equivalent. The ESP32 used the simplified SOC structure before not allowing for Kconfig extensions. This PR aims to change it and follow standardised SOC structuring: Referring the the general soc as `CONFIG_SOC_SERIES_*` and leaving the `CONFIG_SOC_ESP*` for SOC particular model - as it is done in all other recently supported SOCs.

marekmatej commented on dts/xtensa/espressif/esp32s2/esp32s2_common.dtsi:
```diff
@@ -0,0 +1,316 @@
+ * Copyright (c) 2021 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
The role of `esp32*_common` is to provide common ground for other SOC/SIP modules on DT level. Some particular SOCs, like this `esp32s2` for example, are just fine with what is commonly defined, others extend the functionality.

marekmatej commented on boards/xtensa/esp32_devkitc_wroom/doc/index.rst:
```diff
@@ -0,0 +1,203 @@
+.. _esp32_devkitc_wroom:
+
+ESP32-DEVKITC-WROOM
+###################
```
I have considered that approach at the beginning but since that change is tightly bound with the `soc/xtensa/espressif_esp32` and hasn't much meaning without each other, I choose to merge this in the single PR.

marekmatej commented on boards/xtensa/esp32_devkitc_wroom/esp32_devkitc_wroom.dts:
```diff
@@ -0,0 +1,177 @@
+ */
+/dts-v1/;
+
+#include <espressif/esp32/esp32_wroom_32ue_n4.dtsi>
```
Already reordered the `dts` changes before using them in the `board`.

marekmatej commented on boards/xtensa/esp32s2_saola/esp32s2_saola.dts:
```diff
@@ -6,7 +6,7 @@
 /dts-v1/;
 
-#include <espressif/esp32s2.dtsi>
+#include <espressif/esp32s2/esp32s2_wrover_n4r2.dtsi>
```
Already reordered the dts changes before using them in the board.

mniestroj reviewed (changes requested):
Added some comments to previous (above) threads.

marekmatej committed:
boards: xtensa: esp32 board split

Remove virtual esp32 board and replace it with the
real word boards:

- esp32_devkitc_wroom
- esp32_devkitc_wrover (with PSRAM option)

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej commented:
> Added some comments to previous (above) threads.

PTAL

mniestroj commented on arch/xtensa/Kconfig:
```diff
@@ -46,7 +46,7 @@ config XTENSA_USE_CORE_CRT1
 	bool "Backtrace on panic exception"
 	default y
-	depends on SOC_ESP32 || SOC_FAMILY_INTEL_ADSP
+	depends on SOC_SERIES_ESP32 || SOC_FAMILY_INTEL_ADSP
```
`SOC_ESP32` -> `SOC_SERIES_ESP32` happens in `hal_espressif`, so updating west manifest should be atomic (in single commit) with this change.

marekmatej replied:
Updating the `west.yml` is not sufficient since it must be followed by `west update hal_espressif` call. This is true for any other PR involving the hal changes. Please take a look at the history.

mniestroj replied:
The assumption is that `west update` is called whenever you checkout on new commit. So for example when doing `git bisect` you do `west update` every time. Which means... that `west.yml` update needs to be in the same commit as `SOC_ESP32` -> `SOC_SERIES_ESP32` in the Zephyr codebase.

mniestroj commented on soc/xtensa/espressif_esp32/esp32/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,69 @@
+config SOC_SERIES
+	default "esp32"
+
+config SOC_PART_NUMBER
```
@marekmatej Could you please elaborate why this Kconfig option is introduced? I don't see it used anywhere.

mniestroj commented on boards/xtensa/esp32s2_saola/esp32s2_saola.dts:
```diff
@@ -6,7 +6,7 @@
 /dts-v1/;
 
-#include <espressif/esp32s2/esp32s2_wrover_n8r2.dtsi>
+#include <espressif/esp32s2/esp32s2_wrover_n4r2.dtsi>
```
Is it a fixup of previous commit? Which SiP is the right one?

marekmatej replied:
Fixup according to the https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/hw-reference/esp32s2/user-guide-saola-1-v1.2.html#description-of-components

mniestroj replied:
But you first modify this file to:
```
#include <espressif/esp32s2/esp32s2_wrover_n8r2.dtsi>
``` 
in `dts: xtensa: esp32xx rework soc/sip list`, and then to:
```
#include <espressif/esp32s2/esp32s2_wrover_n4r2.dtsi>
```
in `soc: xtensa,riscv: esp32xx: refactor folder structure`.

So I guess everything should be part of  `dts: xtensa: esp32xx rework soc/sip list`, right?

mniestroj commented on boards/xtensa/esp32s2_saola/esp32s2_saola.dts:
```diff
@@ -120,8 +120,11 @@
+
 &flash0 {
 	status = "okay";
+	reg = <0x0 DT_SIZE_M(4)>;
+
```
This is already inherited from `esp32s2_wrover_n4r2.dtsi`

marekmatej replied:
Thanks for pointing this out! Fixed.

mniestroj replied:
Nope, this is still part of `soc: xtensa,riscv: esp32xx: refactor folder structure`.

mniestroj commented on boards/xtensa/esp32s3_devkitm/esp32s3_devkitm.dts:
```diff
@@ -5,7 +5,7 @@
 /dts-v1/;
 
-#include <espressif/esp32s3/esp32s3_wroom_n8r8.dtsi>
+#include <espressif/esp32s3/esp32s3_mini_n8.dtsi>
```
Is it a fixup of previous commit? Which SiP is the right one?

marekmatej replied:
Thanks, fixed!

mniestroj commented on boards/xtensa/heltec_wifi_lora32_v2/heltec_wifi_lora32_v2.dts:
```diff
@@ -5,7 +5,7 @@
 /dts-v1/;
 
-#include <espressif/esp32/esp32_wrover_e_n4r8.dtsi>
+#include <espressif/esp32/esp32_d0wd_v3.dtsi>
```
Is it a fixup of previous commit?

marekmatej replied:
This is fixup since `heltec_wifi_lora32_v2` does not use SIP.

mniestroj commented on boards/xtensa/m5stickc_plus/m5stickc_plus.dts:
```diff
@@ -5,7 +5,7 @@
 /dts-v1/;
 
-#include <espressif/esp32/esp32_wrover_e_n4r8.dtsi>
+#include <espressif/esp32/esp32_pico_d4.dtsi>
```
Is it a fixup of previous commit? Which SiP is the right one?

marekmatej replied:
The last one is correct.

mniestroj commented on boards/xtensa/odroid_go/odroid_go.dts:
```diff
@@ -5,7 +5,7 @@
 /dts-v1/;
 
-#include <espressif/esp32/esp32_wrover_e_n4r8.dtsi>
+#include "espressif/esp32/esp32_wrover_e_n16r2.dtsi"
```
Is it a fixup of previous commit? Which SiP is the right one?
Also: please leave `<>` instead of `""`.

marekmatej replied:
Thanks for the finding! Actually the `odroig_go` uses 4MB os PSRAM so I've added a new dtsi.

mniestroj commented on boards/xtensa/olimex_esp32_evb/olimex_esp32_evb.dts:
```diff
@@ -6,7 +6,7 @@
 /dts-v1/;
 
-#include <espressif/esp32/esp32_wrover_e_n4r8.dtsi>
+#include <espressif/esp32/esp32_wrover_e_n4r2.dtsi>
```
Is it a fixup of previous commit? Which SiP is the right one?

marekmatej replied:
This one is tricky since the manufacturer is not specific about the installed Flash or Spiram. So I leave it as is.

mniestroj commented on boards/riscv/esp32c3_devkitm/Kconfig.board:
```diff
@@ -5,4 +5,8 @@
 config BOARD_ESP32C3_DEVKITM
 	bool "ESP32C3 Devkit-M Board"
-	depends on SOC_ESP32C3
+	depends on SOC_SERIES_ESP32C3
```
Part of those changes (`SOC_ESP32C3` -> `SOC_SERIES_ESP32C3`) were already done in previous commit ("soc: xtensa: esp32xx: refactor folder structure"). Please keep them together.

marekmatej replied:
Ok fixed.

mniestroj reviewed (dismissed):
Main issues:
- `CONFIG_SOC` -> `CONFIG_SOC_SERIES` needs to be done with `west.yml` update to have it in sync with changes in HAL
- some devicetree changes are made twice (to one SiP variant, then to another)
- `SOC_ESP32C3` -> `SOC_SERIES_ESP32C3` is not atomic
- introduction of `CONFIG_SOC_PART_NUMBER` is not justified and this symbol seems not to be used anywhere

marekmatej commented:
> Main issues:
> 
> * `CONFIG_SOC` -> `CONFIG_SOC_SERIES` needs to be done with `west.yml` update to have it in sync with changes in HAL

Yes, part of the changes are in `hal_espressif`, and a manual run of `west update` is needed. As with any other PR, which depends on hal changes. The manifest change commit reflects that.

> * some devicetree changes are made twice (to one SiP variant, then to another)

See particular comments below.

> * `SOC_ESP32C3` -> `SOC_SERIES_ESP32C3` is not atomic

Commits for Xtensa and RISCV architectures were separated deliberately. If you wish so I can squash the 2 commits.

> * introduction of `CONFIG_SOC_PART_NUMBER` is not justified and this symbol seems not to be used anywhere

It is used to provide the SOC name string, which is intended for future use.

mniestroj commented:
> > * `SOC_ESP32C3` -> `SOC_SERIES_ESP32C3` is not atomic
> 
> Commits for Xtensa and RISCV architectures were separated deliberately. If you wish so I can squash the 2 commits.

Ideally there should be a separate (and only one) commit with `s/CONFIG_SOC/CONFIG_SOC_SERIES/` change, for all the ESP32 boards (both xtensa and riscv) in the tree AND `west.yml` (since it assumes that change).  They just cannot be in separate commits, as that would break bisectability (and this is a single logical change anyway).

> 
> > * introduction of `CONFIG_SOC_PART_NUMBER` is not justified and this symbol seems not to be used anywhere
> 
> It is used to provide the SOC name string, which is intended for future use.

So no reason to keep that in single commit with all the other changes. One thing that is suspicious with the commits you created is that you mention several things in a list. Sometimes they need to go together, but in that case I think that stuff like 1) `s/CONFIG_SOC/CONFIG_SOC_SERIES/`, 2) `CONFIG_SOC_PART_NUMBER` introduction, 3) other stuff should all go separately.

Please also provide the reasoning behing `CONFIG_SOC_PART_NUMBER` introduction in the commit message. There needs to be a WHY provided. It might happen in the end that plans change and that you won't be using it. In that case it would be good to know WHY it was introduced in the first place and whether it is subject for removal (as part of refactoring / cleaning up). Another thing is that this probably should be part of separate PR, since introduction of new symbol (which is not even used) is not part of "Refactor ESP32xx folder structure" which this PR is mainly about. The smaller the PRs, the better.

str4t0m commented:
Have you considered adding the renamed boards to the `boards/deprecated.cmake` file?
It would allow to build under the old name while getting a warning that the board was renamed.

marekmatej committed:
boards: mark esp32 board as deprecated

Mark the 'esp32' board as deprecated after removing
it and replaced by the two real boards.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej commented:
> Have you considered adding the renamed boards to the `boards/deprecated.cmake` file? It would allow to build under the old name while getting a warning that the board was renamed.

Thank you @str4t0m for pointing that out! I've updated the PR accordingly.

marekmatej committed:
soc: xtensa,riscv: esp32xx: refactor folder structure

Refactor the ESP32 target SOCs together with
all related boards. Most braking changes includes:

- changing the CONFIG_SOC_ESP32* to refer to
  the actual soc line (esp32,esp32s2,esp32s3,esp32c3)
- replacing CONFIG_SOC with the CONFIG_SOC_SERIES
- creating CONFIG_SOC_FAMILY_ESP32 to embrace all
  the ESP32 across all used architectures
- introducing CONFIG_SOC_PART_NUMBER_* to
  provide a SOC model config
- introducing the 'common' folder to hide all
  commonly used configs and files.
- updating west.yml to reflect previous changes in hal

Signed-off-by: Marek Matej <marek.matej@espressif.com>

mniestroj commented on dts/riscv/espressif/esp32c3/esp32c3_wroom_h2.dtsi:
```diff
@@ -0,0 +1,12 @@
+
+#include "esp32c3_common.dtsi"
+
+/* 4MB flash */
+&flash0 {
+	reg = <0x0 DT_SIZE_M(4)>;
```
Since `esp32c3_wroom_h4` has 4MB flash, doesn't `esp32c3_wroom_h2` contain only 2MB flash?

marekmatej replied:
thanks, fixed

mniestroj reviewed (commented):
Two last comments:
- there is still not a good rationale why `CONFIG_SOC_PART_NUMBER` is introduced, so please remove it for now and introduce it once there will be a use case for it
  - it will make commit diff a bit smaller, since it is not directly related to "refactoring folder structure"
  - this "part number" (which seems to be SoC or SiP name) should probably come from devicetree (e.g. derived from compatible string) instead of being set by Kconfig explicitly (it will make it easier to introduce new boards, since such hardware information will be put only in one place)
- one inline comment about `h2` variant of `esp32c3` SiP

marekmatej commented:
> Two last comments:
> 
> * there is still not a good rationale why `CONFIG_SOC_PART_NUMBER` is introduced, so please remove it for now and introduce it once there will be a use case for it
>   
>   * it will make commit diff a bit smaller, since it is not directly related to "refactoring folder structure"
>   * this "part number" (which seems to be SoC or SiP name) should probably come from devicetree (e.g. derived from compatible string) instead of being set by Kconfig explicitly (it will make it easier to introduce new boards, since such hardware information will be put only in one place)

We decided to use the `SOC_PART_NUMBER` symbol in our port because of the following reasons:
1. the symbol has been introduced a long time ago and it is mentioned in the SOC contributing guideline in the examples
2. many other vendors are having it even without referencing it - check the `soc/arm/atmel_* | nxp_* | cypress | silabs_* | infineon* | ti_*` 
3. it is handy for users as they can use the SOC string for runtime identification of their devices
4. it is not producing any code overhead

mniestroj commented:
> > Two last comments:
> > 
> > * there is still not a good rationale why `CONFIG_SOC_PART_NUMBER` is introduced, so please remove it for now and introduce it once there will be a use case for it
> >   
> >   * it will make commit diff a bit smaller, since it is not directly related to "refactoring folder structure"
> >   * this "part number" (which seems to be SoC or SiP name) should probably come from devicetree (e.g. derived from compatible string) instead of being set by Kconfig explicitly (it will make it easier to introduce new boards, since such hardware information will be put only in one place)
> 
> We decided to use the `SOC_PART_NUMBER` symbol in our port because of the following reasons:
> 
> 1. the symbol has been introduced a long time ago and it is mentioned in the SOC contributing guideline in the examples

In ESP-IDF / HAL?

> 2. many other vendors are having it even without referencing it - check the `soc/arm/atmel_* | nxp_* | cypress | silabs_* | infineon* | ti_*`

Atmel - those symbols are used to select header files.
Cypress, NXP, Infineon - uses it for selecting files in HAL, so there is some use case for it.

In case of Atmel and NXP those symbols started to be used before devicetree was so widespread. I guess that now it would be better to generate those (even if at Kconfig level, but at least automatically/implicitly) from devicetree.

> 3. it is handy for users as they can use the SOC string for runtime identification of their devices

There are `/model` and `/compatible` in root devicetree node, based on which such message can be generated. Let's not introduce another place (Kconfig) which describes hardware model.

> 4. it is not producing any code overhead

If you mean flash footprint - agree. But it takes code in terms of Kconfig lines, which adds maintenance. As I said previously, this unnecessary adds another step in adding support for custom ESP32-based boards. It should be devicetree where SoC/SiP information is provided. If this information will be useful at Kconfig level, then a glue code based on Kconfig functions could be developed to set that automatically for each board (based on devicetree).

I am not convinced by above reasoning. And once again... introduction of `CONFIG_SOC_PART_NUMBER` has nothing to do with "refactor folder structure" commit.

marekmatej commented:
> In ESP-IDF / HAL?

In Zephyr board contribution guideline.

marekmatej commented:
> If you mean flash footprint - agree. But it takes code in terms of Kconfig lines, which adds maintenance. As I said previously, this unnecessary adds another step in adding support for custom ESP32-based boards. It should be devicetree where SoC/SiP information is provided. If this information will be useful at Kconfig level, then a glue code based on Kconfig functions could be developed to set that automatically for each board (based on devicetree).

That is actually not true as custom boards could use plain `SOC_ESP33xx` and specify their memories anywhere in the `dts`.

> I am not convinced by above reasoning. And once again... introduction of `CONFIG_SOC_PART_NUMBER` has nothing to do with "refactor folder structure" commit.

I'm thankful for your in-depth review, but I'll let the TSC decide whether this should be removed or not. I already addressed all your valid change requests, and this one doesn't seem to be justified enough.

sylvioalves commented:
@henrikbrixandersen @nashif Hi, would you be able to take a look at this?

carlescufi commented:
> > In ESP-IDF / HAL?
> 
> In Zephyr board contribution guideline.

I agree with @marekmatej, this is directly in our doc @mniestroj:

https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#write-kconfig-files

```
config BOARD_PLANK
   bool "Plank board"
   depends on SOC_SERIES_YOUR_SOC_SERIES_HERE
   select SOC_PART_NUMBER_ABCDEFGH
```

henrikbrixandersen commented:
> I agree with @marekmatej, this is directly in our doc @mniestroj:
> 
> https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#write-kconfig-files
> 
> ```
> config BOARD_PLANK
>    bool "Plank board"
>    depends on SOC_SERIES_YOUR_SOC_SERIES_HERE
>    select SOC_PART_NUMBER_ABCDEFGH
> ```

I agree as well. I see no issue with introducing `SOC_PART_NUMBER_*` Kconfigs for ESP32xx, even if they are unused in upstream Zephyr/HAL.

We use `SOC_PART_NUMBER_*` for conditional compilation in our downstream applications for other SoCs.

henrikbrixandersen commented:
@marekmatej FYI: This needs rebasing.

gmarull commented:
> > > In ESP-IDF / HAL?
> > 
> > 
> > In Zephyr board contribution guideline.
> 
> I agree with @marekmatej, this is directly in our doc @mniestroj:
> 
> https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#write-kconfig-files
> 
> ```
> config BOARD_PLANK
>    bool "Plank board"
>    depends on SOC_SERIES_YOUR_SOC_SERIES_HERE
>    select SOC_PART_NUMBER_ABCDEFGH
> ```

I think the question is: where is this Kconfig option used? If not used please, just drop it regardless of what porting guidelines say. Guidelines contain examples that may not always apply.

carlescufi commented:
> > > > In ESP-IDF / HAL?
> > > 
> > > 
> > > In Zephyr board contribution guideline.
> > 
> > 
> > I agree with @marekmatej, this is directly in our doc @mniestroj:
> > https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#write-kconfig-files
> > ```
> > config BOARD_PLANK
> >    bool "Plank board"
> >    depends on SOC_SERIES_YOUR_SOC_SERIES_HERE
> >    select SOC_PART_NUMBER_ABCDEFGH
> > ```
> 
> I think the question is: where is this Kconfig option used? If not used please, just drop it regardless of what porting guidelines say. Guidelines contain examples that may not always apply.

But what if downstream users want to make use of it? I think it's legitimate for a vendor to provide users with a string representation of the part number for downstream use.

fabiobaltieri commented:
> I think the question is: where is this Kconfig option used? If not used please, just drop it regardless of what porting guidelines say. Guidelines contain examples that may not always apply.

We are talking about `config SOC_PART_NUMBER` right? The string one. If there's no solid use case we may as well drop it everywhere, one less thing to care about when porting. The boolean `CONFIG_SOC_PART_NUMBER_...` ones seems to be used and makes more sense.

fabiobaltieri commented:
> For string representations, we already have `CONFIG_SOC_SERIES` (e.g. stm32f4), `CONFIG_SOC` (e.g. `stm32f401xc`) and some `CONFIG_SOC_FAMILY`, so I'm not convinced why we need yet another symbol. As I said, the question is why do we need those symbols.

These are used in `soc.cmake` to build up `SOC_PATH`, maybe `SOC_PART_NUMBER` was added as a logical extension of that but if it's not needed it's not needed.

henrikbrixandersen commented:
> We are talking about `config SOC_PART_NUMBER` right? The string one. If there's no solid use case we may as well drop it everywhere, one less thing to care about when porting. The boolean `CONFIG_SOC_PART_NUMBER_...` ones seems to be used and makes more sense.

Thank you for clearing that up. I was under the impression we were discussing the boolean `CONFIG_SOC_PART_NUMBER_...`, which I do find useful - and which are used in our downstream.

I agree that the string representation seem unused and superfluous.

fabiobaltieri commented:
> I agree that the string representation seem unused and superfluous.

Yeah it's all a bit confusing, best as I can tell `config SOC_PART_NUMBER` is also not defined at top level but only by individual archs, won't go through the code archeology maybe it was used for something specific historically and now it's not anymore? There's a reference in `modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt` and that's it.

marekmatej commented:
Ok, so those `CONFIG_SOC_PART_NUMBER` strings were added for downstream users so they can be used to identify the underlying hw. There is no other meaningful usage, except maybe creating the sample to code to use those strings as @carlescufi noted, but that would be very artificial justification.
So I'll remove those strings from the PR. Thank you all for your clarification.

mniestroj commented on soc/riscv/espressif_esp32/Kconfig:
```diff
@@ -0,0 +1,23 @@
+	string
+	default "espressif_esp32"
+
+config SOC_PART_NUMBER
+	string
+	help
+	  This string holds the full part number of the SoC. It is a hidden option
+	  that you should not set directly. The part number selection choice defines
+	  the default value for this string.
```
@marekmatej This "string" option should be removed, right?

MaureenHelm commented:
> Ok, so those `CONFIG_SOC_PART_NUMBER` strings were added for downstream users so they can be used to identify the underlying hw. There is no other meaningful usage, except maybe creating the sample to code to use those strings as @carlescufi noted, but that would be very artificial justification. So I'll remove those strings from the PR. Thank you all for your clarification.

I introduced `CONFIG_SOC_PART_NUMBER` many years ago to help integrate the Kinetis SDK, now called MCUXpresso SDK, into Zephyr. See commit 281568718687bab84224272815844a9f14dda60c. The pattern has since been copy-pasted to other SoC families.

marekmatej commented:
@mniestroj, at the end I think we were mixing two different things - `SOC_PART_NUMBER` strings and bool flags. I was advocating for both! From your messages, I got the impression you wanted to remove both strings and flags. However, I must admit that having those strings for future use is not vindicated enough, so I apologize for the misunderstanding from my side.

carlescufi commented:
> This PR looks fine to me. I just have issues with `SOC_PART_NUMBER` (not critical though). Since others are good with it (based on the comments above) and overall this PR is bringing lots of net profit (thanks @marekmatej), then I will not stay in the way.

Just to be clear @mniestroj, `SOC_PART_NUMBER` has now been removed from this PR.

mniestroj commented:
> > This PR looks fine to me. I just have issues with `SOC_PART_NUMBER` (not critical though). Since others are good with it (based on the comments above) and overall this PR is bringing lots of net profit (thanks @marekmatej), then I will not stay in the way.
> 
> Just to be clear @mniestroj, `SOC_PART_NUMBER` has now been removed from this PR.

Now we got confused... in this context I meant `choice SOC_PART_NUMBER` with `bool SOC_PART_NUMBER_*`.

`string SOC_PART_NUMBER` was removed, indeed. @carlescufi

carlescufi commented:
> > > This PR looks fine to me. I just have issues with `SOC_PART_NUMBER` (not critical though). Since others are good with it (based on the comments above) and overall this PR is bringing lots of net profit (thanks @marekmatej), then I will not stay in the way.
> > 
> > 
> > Just to be clear @mniestroj, `SOC_PART_NUMBER` has now been removed from this PR.
> 
> Now we got confused... in this context I meant `choice SOC_PART_NUMBER` with `bool SOC_PART_NUMBER_*`.
> 
> `string SOC_PART_NUMBER` was removed, indeed. @carlescufi

Aah, of course. The choice vs the string Kconfig :) But as you say others chimed in in favor of keeping the choice, so thanks for unblocking. Merging this now.
